clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -201,13 +207,20 @@ public int compare(TxnRequestHandler o1, TxnRequestHandler o2) {
 
     public synchronized TransactionalRequestResult initializeTransactions() {
         ensureTransactional();
-        transitionTo(State.INITIALIZING);
-        setProducerIdAndEpoch(ProducerIdAndEpoch.NONE);
-        this.nextSequence.clear();
-        InitProducerIdRequest.Builder builder = new InitProducerIdRequest.Builder(transactionalId, transactionTimeoutMs);
-        InitProducerIdHandler handler = new InitProducerIdHandler(builder);
-        enqueueRequest(handler);
-        return handler.result;
+
+        TransactionalRequestResult initTransactionResult = this.stateToTransactionRequestResult.get(State.INITIALIZING);
```

Comment:
If we get rid of the `stateToTransactionRequestResult` map and just use a single field to track the current pending operation, then this would become a simple check on the current state.

Reply:
Thanks for the comments. If a single field is employed, nulling out that field seems needing to be placed in `TransactionManager.await**` which will not be called by TransactionManagerTest at all. That means, we have to create a clear function `transactionManager.clearCache()` and call it in every method in TransactionManagerTest. Does it make sense?

Reply:
@huxihx I'm not 100% sure I follow. Here is a quick sketch of what I had in mind: https://github.com/hachikuji/kafka/commit/b3364035e4d06defd9a49ddd7c158a4f2721deab. I'm not sure we need any changes to reset the pending operation, at least not for the purpose of the test case. Probably it makes sense to clear it after a state change though.